modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/MergeJoinNode.java

```diff
@@ -1006,27 +943,27 @@ else if (cmp > 0) {
                 inLoop = false;
             }
 
-            if (waitingRight == 0)
-                rightSource().request(waitingRight = IN_BUFFER_SIZE);
-
-            if (waitingLeft == 0)
-                leftSource().request(waitingLeft = IN_BUFFER_SIZE);
-
             if (requested > 0 && waitingLeft == NOT_WAITING && left == null && leftInBuf.isEmpty()
                 && waitingRight == NOT_WAITING && right == null && rightInBuf.isEmpty() && rightMaterialization == null
+                && checkFinished(0, true)
             )
-                checkJoinFinished();
+                return;
+
+            tryToRequestInputs();
         }
     }
 
     /** */
-    private static class SemiJoin<Row> extends MergeJoinNode<Row> {
-        /** */
-        private Row left;
+    protected void tryToRequestInputs() throws Exception {
+        if (waitingLeft == 0 && requested > 0 && leftInBuf.size() <= HALF_BUF_SIZE)
```

Comment:
Why do we need `requested > 0` here? Condition with `leftInBuf.size()` already covers all problems related to buffer overflow. Any tests that can prove that additional condition is required? IMO this condition can add another problems (for example, we need to drain remote sources even if requested is 0).